File: src/test/java/com/github/kayjamlang/tests/expressions/ifExpr/IfWithGetOperationTest.java
package com.github.kayjamlang.tests.expressions.ifExpr;

import com.github.kayjamlang.core.expressions.Expression;
import com.github.kayjamlang.core.KayJamLexer;
import com.github.kayjamlang.core.KayJamParser;
import com.github.kayjamlang.core.expressions.ValueExpression;
import com.github.kayjamlang.core.expressions.GetExpression;
import com.github.kayjamlang.core.expressions.IfExpression;
import com.github.kayjamlang.core.expressions.OperationExpression;
import org.junit.BeforeClass;
import org.junit.Test;

import static org.junit.Assert.*;

public class IfWithGetOperationTest {

    private static KayJamParser parser;

    @BeforeClass
    public static void prepare(){
        parser = new KayJamParser(new KayJamLexer("if(test[1]==123) true"));
    }

    @Test
    public void test() throws Exception {
        Expression expression = parser.readExpression();

        assertNotNull(expression);
        assertSame(IfExpression.class, expression.getClass());

        IfExpression ifExpression = (IfExpression) expression;
        assertSame(OperationExpression.class, ifExpression.condition.getClass());

        assertNotNull(ifExpression.ifTrue);
        assertNull(ifExpression.ifFalse);

        assertSame(ValueExpression.class, ifExpression.ifTrue.getClass());

        OperationExpression root = (OperationExpression) ifExpression.condition;
        assertSame(GetExpression.class, root.left.getClass());
        assertSame(ValueExpression.class, root.right.getClass());

        ValueExpression ifTrue = (ValueExpression) ifExpression.ifTrue;
        assertEquals(true, ifTrue.value);
    }
}

File: src/test/scala/com/github/kayjamlang/tests/expressions/ifExpr/IfWithGetOperationTest.scala
package com.github.kayjamlang.tests.expressions.ifExpr

import com.github.kayjamlang.core.KayJamLexer
import com.github.kayjamlang.core.KayJamParser
import com.github.kayjamlang.core.expressions._
import org.junit.BeforeClass
import org.junit.Test
import org.junit.Assert._


object IfWithGetOperationTest {
  private var parser: KayJamParser = null

  @BeforeClass def prepare(): Unit = {
    parser = new KayJamParser(new KayJamLexer(
      """
        |if (test[1] == 123)
        | true
        |""".stripMargin))
  }
}

class IfWithGetOperationTest {
  @Test
  @throws[Exception]
  def test(): Unit = {
    val expression = IfWithGetOperationTest.parser.readExpression
    assertNotNull(expression)
    assertSame(classOf[IfExpression], expression getClass)
    val ifExpression = expression.asInstanceOf[IfExpression]
    assertSame(classOf[OperationExpression], ifExpression.condition getClass)
    assertNotNull(ifExpression.ifTrue)
    assertNull(ifExpression.ifFalse)
    assertSame(classOf[ValueExpression], ifExpression.ifTrue getClass)
    val root = ifExpression.condition.asInstanceOf[OperationExpression]
    assertSame(classOf[GetExpression], root.left getClass)
    assertSame(classOf[ValueExpression], root.right getClass)
    val ifTrue = ifExpression.ifTrue.asInstanceOf[ValueExpression]
    assertEquals(true, ifTrue.value)
  }
}